dbms/src/Flash/Mpp/AsyncRequestHandler.h

```diff
@@ -0,0 +1,389 @@
+    using AsyncReader = typename RPCContext::AsyncReader;
+    using Self = AsyncRequestHandler<RPCContext, enable_fine_grained_shuffle>;
+
+    AsyncRequestHandler(
+        std::vector<GRPCReceiveQueue<RecvMsgPtr>> & grpc_recv_queues_,
+        AsyncRequestHandlerWaitQueuePtr async_wait_rewrite_queue_,
+        const std::shared_ptr<RPCContext> & context,
+        Request && req,
+        const String & req_id,
+        std::atomic<Int64> * data_size_in_queue,
+        std::function<void(bool, const String &, const LoggerPtr &)> && close_conn_)
+        : cq(&(GRPCCompletionQueuePool::global_instance->pickQueue()))
+        , rpc_context(context)
+        , request(std::move(req))
+        , req_info(fmt::format("tunnel{}+{}", req.send_task_id, req.recv_task_id))
+        , has_data(false)
+        , retry_times(0)
+        , stage(AsyncRequestStage::NEED_INIT)
+        , read_packet_index(0)
+        , received_packet_index(0)
+        , finish_status(RPCContext::getStatusOK())
+        , log(Logger::get(req_id, req_info))
+        , channel_try_writer(grpc_recv_queues_, req_info, log, data_size_in_queue, ReceiverMode::Async)
+        , async_wait_rewrite_queue(std::move(async_wait_rewrite_queue_))
+        , kick_recv_tag(thisAsUnaryCallback())
+        , close_conn(std::move(close_conn_))
+        , is_close_conn_called(false)
+    {
+        packets.resize(batch_packet_count);
+        for (auto & packet : packets)
+            packet = std::make_shared<TrackedMppDataPacket>(MPPDataPacketV0);
+        // TODO add random fail point to mock the fail
+        start();
+    }
+
+    // execute will be called by RPC framework so it should be as light as possible.
+    // Do not do anything after processXXX functions.
+    void execute(bool & ok) override
+    {
+        try
+        {
+            switch (stage)
+            {
+            case AsyncRequestStage::WAIT_RETRY:
+                start();
+                break;
+            case AsyncRequestStage::WAIT_MAKE_READER:
+                processWaitMakeReader(ok);
+                break;
+            case AsyncRequestStage::WAIT_BATCH_READ:
+                processWaitBatchRead(ok);
+                break;
+            case AsyncRequestStage::WAIT_REWRITE:
+                processWaitReWrite();
+                break;
+            case AsyncRequestStage::WAIT_FINISH:
+                processWaitFinish();
+                break;
+            default:
+                RUNTIME_ASSERT(false, "Unexpected stage {}", magic_enum::enum_name(stage));
+            }
+        }
+        catch (...)
+        {
+            LOG_ERROR(log, "Some errors happen in the AsyncRequestHandler.execute()");
+            closeConnection("Exception is thrown in AsyncRequestHandler");
+        }
+    }
+
+private:
+    void processWaitMakeReader(bool ok)
+    {
+        // Use lock to ensure reader is created already in reactor thread
+        std::lock_guard lock(make_reader_mu);
```

Comment:
Will this lock block block for a long time?
Is it necessary to use a non-blocking way of checking?
```
if !(is_ready)
    return;
```

Reply:
> Will this lock block block for a long time? Is it necessary to use a non-blocking way of checking?
> 
> ```
>     void start()
>     {
>         stage = AsyncRequestStage::WAIT_MAKE_READER;
> 
>         rpc_context->makeAsyncReader(request, reader, cq, thisAsUnaryCallback());
>         is_reader_ready = true; // std::atomic_bool
>     }
> ```
> 
> ```
>     void processWaitMakeReader(bool ok)
>     {
>         // Use lock to ensure reader is created already in reactor thread
>         std::lock_guard lock(make_reader_mu);
>         if (!ok)
>         {
>             reader.reset();
>             LOG_WARNING(log, "MakeReader fail. retry time: {}", retry_times);
>             retryOrDone("Exchange receiver meet error : send async stream request fail");
>         }
>         else
>         {
>             if (!is_reader_ready)
>                 return;
>             startAsyncRead();
>         }
>     }
> ```

First, data race here is very hard to trigger. Second, only when failing to creating connection, the data race will happen, and performance is unnecessary to focus on in this circumstance.

Reply:
> > Will this lock block block for a long time? Is it necessary to use a non-blocking way of checking?
> > ```
> >     void start()
> >     {
> >         stage = AsyncRequestStage::WAIT_MAKE_READER;
> > 
> >         rpc_context->makeAsyncReader(request, reader, cq, thisAsUnaryCallback());
> >         is_reader_ready = true; // std::atomic_bool
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> >     void processWaitMakeReader(bool ok)
> >     {
> >         // Use lock to ensure reader is created already in reactor thread
> >         std::lock_guard lock(make_reader_mu);
> >         if (!ok)
> >         {
> >             reader.reset();
> >             LOG_WARNING(log, "MakeReader fail. retry time: {}", retry_times);
> >             retryOrDone("Exchange receiver meet error : send async stream request fail");
> >         }
> >         else
> >         {
> >             if (!is_reader_ready)
> >                 return;
> >             startAsyncRead();
> >         }
> >     }
> > ```
> 
> First, data race here is very hard to trigger. Second, only when failing to creat connection, the data race will happen, and performance is unnecessary to focus on in this circumstance.

Reply:
ok